src/framework/xr/xr-view.js

```diff
@@ -0,0 +1,338 @@
+     * @param {XRView} xrView - XRView from WebXR API.
+     * @ignore
+     */
+    update(frame, xrView) {
+        this._xrView = xrView;
+        if (this._manager.views.availableColor)
+            this._xrCamera = this._xrView.camera;
+
+        const layer = frame.session.renderState.baseLayer;
+
+        // viewport
+        const viewport = layer.getViewport(this._xrView);
+        this._viewport.x = viewport.x;
+        this._viewport.y = viewport.y;
+        this._viewport.z = viewport.width;
+        this._viewport.w = viewport.height;
+
+        // matrices
+        this._projMat.set(this._xrView.projectionMatrix);
+        this._viewMat.set(this._xrView.transform.inverse.matrix);
+        this._viewInvMat.set(this._xrView.transform.matrix);
+
+        this._updateTextureColor();
+    }
+
+    /**
+     * @private
+     */
+    _updateTextureColor() {
+        if (!this._manager.views.availableColor || !this._xrCamera)
+            return;
+
+        const binding = this._manager.webglBinding;
+        if (!binding)
+            return;
+
+        const texture = binding.getCameraImage(this._xrCamera);
+        if (!texture)
+            return;
+
+        const device = this._manager.app.graphicsDevice;
+        const gl = device.gl;
+        const attachmentBaseConstant = device.isWebGL2 ? gl.COLOR_ATTACHMENT0 : (device.extDrawBuffers?.COLOR_ATTACHMENT0_WEBGL ?? gl.COLOR_ATTACHMENT0);
+
+        const width = this._xrCamera.width;
+        const height = this._xrCamera.height;
+
+        if (!this._textureColor) {
+            // color texture
+            this._textureColor = new Texture(this._manager.app.graphicsDevice, {
+                format: PIXELFORMAT_RGB8,
+                mipmaps: false,
+                addressU: ADDRESS_CLAMP_TO_EDGE,
+                addressV: ADDRESS_CLAMP_TO_EDGE,
+                minFilter: FILTER_LINEAR,
+                magFilter: FILTER_LINEAR,
+                width: width,
+                height: height,
+                name: `XrView-${this._xrView.eye}-Color`
+            });
+
+            // force initialize texture
+            this._textureColor.upload();
+
+            // create frame buffer to read from
+            this._frameBufferSource = gl.createFramebuffer();
+
+            // create frame buffer to write to
+            this._frameBuffer = gl.createFramebuffer();
+        }
+
+        // set frame buffer to read from
+        device.setFramebuffer(this._frameBufferSource);
+        gl.framebufferTexture2D(
+            gl.FRAMEBUFFER,
+            attachmentBaseConstant,
+            gl.TEXTURE_2D,
+            texture,
+            0
+        );
+
+        // set frame buffer to write to
+        device.setFramebuffer(this._frameBuffer);
+        gl.framebufferTexture2D(
+            gl.FRAMEBUFFER,
+            attachmentBaseConstant,
+            gl.TEXTURE_2D,
+            this._textureColor.impl._glTexture,
+            0
+        );
+
+        // bind buffers
```

Comment:
I'd recommend:
- call checkFramebufferStatus in debug mode only (the rest of the engine does `Debug.call(() => this._checkFbo(device, target));` - but only when the framebuffer is created, not each frame, as this is expensive call.

Reply:
Good tip. I've noticed that frame buffer were not ready to be used straight away, how to check/skip the `blit` without checks?

Reply:
we don't do any checks like this in the engine .. the buffer should be ready straight away, so that's something strange going on there. Do you get any webgl errors?

Reply:
Yes, I get this error:
```
[.WebGL-0x200cb2200] GL_INVALID_FRAMEBUFFER_OPERATION: Framebuffer is incomplete: No attachments and default size is zero.
```

Reply:
so that feels like maybe the texture from `texture = binding.getCameraImage` is zero size initially or something .. maybe there's a way to test for this instead and early out, same as when it's null

Reply:
From my observation, Android Chrome likely has multiple textures behind the scenes, where they provide swizzled access to. As they likely render into one texture while providing access to other. That is partly why they are "opaque", so we don't try to read texture which they might be writing to.

I found a workaround. Basically do not try to copy data on the first frame of session, but do it on the second - that worked well. I do create a texture on initialize to have more stable behavour for the user.